### ISSUE TITLE
Build errors with Boost 1.83

### ISSUE THREAD

**Issue by jkhsjdhjs** [p4lang/p4c#4147]
Hey,

since upgrading to Boost 1.83 I can't build the p4c graphs backend anymore due to the following error:
```
In file included from /usr/include/boost/multi_index/hashed_index.hpp:32,
                 from /usr/include/boost/graph/named_graph.hpp:18,
                 from /usr/include/boost/graph/adjacency_list.hpp:36,
                 from /home/jkhsjdhjs/aur/p4lang-p4c/src/p4c-1.2.4.3/backends/graphs/graphs.h:33,
                 from /home/jkhsjdhjs/aur/p4lang-p4c/src/p4c-1.2.4.3/backends/graphs/graphs.cpp:17,
                 from /home/jkhsjdhjs/aur/p4lang-p4c/src/p4c-1.2.4.3/build/backends/graphs/CMakeFiles/p4cgraphs.dir/Unity/unity_0_cxx.cxx:4:
/usr/include/boost/multi_index/detail/bucket_array.hpp: In function ‘void boost::serialization::load_construct_data(Archive&, boost::multi_index::detail::bucket_array<Allocator>*, unsigned int)’:
/usr/include/boost/multi_index/detail/bucket_array.hpp:239:19: error: there are no arguments to ‘bad_archive_exception’ that depend on a template parameter, so a declaration of ‘bad_archive_exception’ must be available [-fpermissive]
  239 |   throw_exception(bad_archive_exception());
      |                   ^~~~~~~~~~~~~~~~~~~~~
/usr/include/boost/multi_index/detail/bucket_array.hpp:239:19: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
In file included from /usr/include/boost/multi_index/hashed_index.hpp:35:
/usr/include/boost/multi_index/detail/index_node_base.hpp: In function ‘void boost::serialization::load_construct_data(Archive&, boost::multi_index::detail::index_node_base<Value, Allocator>*, unsigned int)’:
/usr/include/boost/multi_index/detail/index_node_base.hpp:120:19: error: there are no arguments to ‘bad_archive_exception’ that depend on a template parameter, so a declaration of ‘bad_archive_exception’ must be available [-fpermissive]
  120 |   throw_exception(bad_archive_exception());
      |                   ^~~~~~~~~~~~~~~~~~~~~
```

When building without the graphs backend (`-DENABLE_P4C_GRAPHS=OFF`) it works fine, just like before. I've previously built the p4c graphs backend with Boost 1.81, which also worked fine.
Since I'm not sure whether this is a Boost bug or an incompatibility of p4c with Boost 1.83, I hope that you can give me a few pointers here. Thanks in advance!

**Comment by fruffy** [comment]
Boost is one of these dependencies that constantly introduces backwards-incompatible changes. It does not look like it is the graph's back ends fault? 

It might also be a conflict between to different versions of boost, e.g., `graph` is `1.81` and `multi_index` is 1.83 or something.

**Comment by jkhsjdhjs** [comment]
Hmm, I definitely have the same version of `graph` and `multi_index`, since on Arch there is only one package providing all headers and another package providing all runtime libraries and I currently have version 1.83 installed of both.
I forwarded this to boost: https://github.com/boostorg/multi_index/issues/70

**Comment by jkhsjdhjs** [comment]
Alright, the issue got resolved pretty quickly. It was indeed a bug in Boost.MultiIndex that is triggered when `-DBOOST_NO_ARGUMENT_DEPENDENT_LOOKUP` is used. This flag declares that the compiler doesn't support argument dependent lookups, which only applies to really old compilers. However, it is set for all builds in the CMakeLists.txt:
https://github.com/p4lang/p4c/blob/87d362f409685159470628d565b430abe3342300/CMakeLists.txt#L150-L151

So while the bug is fixed with Boost 1.84, maybe this flag shouldn't be used anymore.

**Comment by fruffy** [comment]
Thanks for investigating this! If we are able to remove it that would be great. I think the compiler did something funky with overloading of ordered_map, which necessitated this definition.  

@ChrisDodd Do you have any issues with removing `-DBOOST_NO_ARGUMENT_DEPENDENT_LOOKUP`?
